File: frontend/src/types/video_sources.ts
/**
 * Video source types for multi-platform video source integration
 * Feature: 007-multi-platform-video-source-integration
 */

/** Supported video source types */
export type SourceType =
  | 'youtube'
  | 'vimeo'
  | 'dailymotion'
  | 'twitch'
  | 'direct'
  | 'hls'
  | 'dash'
  | 'google_drive'
  | 'dropbox'
  | 'onedrive'
  | 'rss_feed'
  | 'unknown';

/** Playback status for a video source */
export type VideoSourceStatus = 'playing' | 'queued' | 'error' | 'completed';

/** Request to detect video source type from URL */
export interface VideoSourceDetectRequest {
  /** URL to detect */
  url: string;
}

/** Response from video source detection */
export interface VideoSourceDetectResponse {
  /** Whether the URL is valid */
  valid: boolean;
  /** Detected source type */
  source_type: SourceType;
  /** Human-readable source type label */
  source_type_label: string;
  /** Extracted metadata (video IDs, file IDs, etc.) */
  metadata: Record<string, unknown>;
  /** Normalized URL */
  normalized_url: string;
  /** Error message if detection failed */
  error?: string;
}

/** Request to validate video source URL */
export interface VideoSourceValidateRequest {
  /** URL to validate */
  url: string;
  /** Whether to check availability (makes HTTP requests) */
  check_availability?: boolean;
}

/** Response from video source validation */
export interface VideoSourceValidateResponse {
  /** Whether the URL is valid */
  valid: boolean;
  /** Detected source type */
  source_type: SourceType;
  /** Human-readable source type label */
  source_type_label: string;
  /** Whether the video is available (if check_availability was true) */
  is_available?: boolean;
  /** List of compatibility issues (e.g., codec incompatibility) */
  compatibility_issues: string[];
  /** Error message if validation failed */
  error?: string;
}

/** Information about a supported video source */
export interface SupportedSourceInfo {
  /** Source type identifier */
  type: SourceType;
  /** Human-readable label */
  label: string;
  /** Description of the source type */
  description: string;
  /** Example URLs */
  examples: string[];
}

/** Response listing all supported video sources */
export interface SupportedSourcesResponse {
  /** List of supported sources */
  sources: SupportedSourceInfo[];
  /** Total count of supported sources */
  total_count: number;
}

/** Video source item (playlist entry or standalone video) */
export interface VideoSource {
  /** Unique identifier */
  id: string;
  /** Channel ID (if assigned to a channel) */
  channel_id?: string;
  /** Stream ID (Clean Architecture) */
  stream_id?: string;
  /** Video URL */
  url: string;
  /** Video title */
  title?: string;
  /** Source type */
  type: SourceType;
  /** Playback status */
  status?: VideoSourceStatus;
  /** Duration in seconds (null if unknown) */
  duration?: number | null;
  /** URL to video thumbnail image */
  thumbnail_url?: string;
  /** Platform-specific metadata (video IDs, channel info, etc.) */
  source_metadata?: Record<string, unknown> | null;
  /** Whether this is a live stream (HLS/DASH) */
  is_live?: boolean;
  /** Whether authentication is needed for cloud storage */
  requires_auth?: boolean;
  /** Encrypted token for cloud storage access */
  auth_token?: string;
  /** Preferred video quality (e.g., '1080p', '720p') */
  quality?: string;
  /** Position in queue */
  position?: number;
  /** ID of user who created this entry */
  created_by?: string;
  /** Creation timestamp */
  created_at: string;
  /** File ID (for local files) */
  file_id?: string;
}

/** Request to create a video source */
export interface VideoSourceCreate {
  /** Video URL */
  url: string;
  /** Video title (optional, will be fetched if not provided) */
  title?: string;
  /** Source type (optional, will be auto-detected if not provided) */
  type?: SourceType;
  /** Preferred video quality */
  quality?: string;
  /** Position in queue */
  position?: number;
}

/** Request to update a video source */
export interface VideoSourceUpdate {
  /** Video URL */
  url?: string;
  /** Video title */
  title?: string;
  /** Source type */
  type?: SourceType;
  /** Playback status */
  status?: VideoSourceStatus;
  /** Duration in seconds */
  duration?: number;
  /** Thumbnail URL */
  thumbnail_url?: string;
  /** Platform-specific metadata */
  source_metadata?: Record<string, unknown>;
  /** Whether this is a live stream */
  is_live?: boolean;
  /** Whether authentication is needed */
  requires_auth?: boolean;
  /** Auth token for cloud storage */
  auth_token?: string;
  /** Preferred video quality */
  quality?: string;
  /** Position in queue */
  position?: number;
}

/** Metadata extracted from video source */
export interface VideoSourceMetadata {
  /** Video ID (platform-specific) */
  video_id?: string;
  /** Playlist ID (for playlists) */
  playlist_id?: string;
  /** Channel ID (for channels) */
  channel_id?: string;
  /** File extension (for direct URLs) */
  extension?: string;
  /** MIME type */
  mime_type?: string;
  /** File size in bytes (for direct URLs) */
  file_size?: number;
  /** Additional platform-specific data */
  additional_data?: Record<string, unknown>;
}

/** Cloud storage authentication info */
export interface CloudStorageAuth {
  /** Type of cloud storage */
  provider: 'google_drive' | 'dropbox' | 'onedrive';
  /** Access token */
  access_token: string;
  /** Refresh token (if available) */
  refresh_token?: string;
  /** Token expiration timestamp */
  expires_at?: string;
}

File: backend/src/api/video_sources.py
"""
Video sources API router for the Telegram broadcast platform.

This module provides REST API endpoints for video source management,
including automatic URL detection, validation, and source type information.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, ConfigDict, field_validator

from src.database import get_db
from src.models.user import User
from src.lib.source_detector import SourceDetector, SourceType
from api.auth import get_current_user


router = APIRouter()


# ============================================================================
# Pydantic Models
# ============================================================================

class VideoSourceDetectRequest(BaseModel):
    """Request model for video source detection."""
    url: str

    @field_validator('url')
    @classmethod
    def validate_url_not_empty(cls, v: str) -> str:
        """Validate that URL is not empty."""
        if not v or not v.strip():
            raise ValueError('URL cannot be empty')
        return v.strip()


class VideoSourceDetectResponse(BaseModel):
    """Response model for video source detection."""
    valid: bool
    source_type: str
    source_type_label: str
    metadata: Dict[str, Any]
    normalized_url: str
    error: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class VideoSourceValidateRequest(BaseModel):
    """Request model for video source validation."""
    url: str
    check_availability: bool = False

    @field_validator('url')
    @classmethod
    def validate_url_not_empty(cls, v: str) -> str:
        """Validate that URL is not empty."""
        if not v or not v.strip():
            raise ValueError('URL cannot be empty')
        return v.strip()


class VideoSourceValidateResponse(BaseModel):
    """Response model for video source validation."""
    valid: bool
    source_type: str
    source_type_label: str
    is_available: Optional[bool] = None
    compatibility_issues: List[str] = []
    error: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class SupportedSourcesResponse(BaseModel):
    """Response model for supported sources list."""
    sources: List[Dict[str, Any]]
    total_count: int


# ============================================================================
# Endpoints
# ============================================================================

@router.post("/detect", response_model=VideoSourceDetectResponse)
async def detect_video_source(
    request: VideoSourceDetectRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Detect video source type from URL.

    Automatically detects the type of video source from a given URL
    and extracts relevant metadata such as video IDs, file IDs, etc.

    Supports:
    - YouTube (videos, playlists, channels)
    - Vimeo (videos, playlists)
    - Twitch (channels, VODs, clips)
    - Dailymotion (videos)
    - Direct video URLs (MP4, WebM, MKV, AVI, etc.)
    - HLS/DASH streaming URLs (.m3u8, .mpd)
    - Cloud storage (Google Drive, Dropbox, OneDrive)
    - RSS/Atom feeds with video enclosures
    """
    try:
        # Normalize the URL
        normalized_url = SourceDetector.normalize_url(request.url)

        # Detect source type
        detection_result = SourceDetector.detect_source(normalized_url)

        # Get human-readable label for source type
        source_type_label = _get_source_type_label(detection_result["source_type"])

        return VideoSourceDetectResponse(
            valid=detection_result["valid"],
            source_type=detection_result["source_type"],
            source_type_label=source_type_label,
            metadata=detection_result["metadata"],
            normalized_url=normalized_url,
            error=detection_result.get("error")
        )

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error detecting video source: {str(e)}"
        )


@router.post("/validate", response_model=VideoSourceValidateResponse)
async def validate_video_source(
    request: VideoSourceValidateRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Validate video source URL and check compatibility.

    Validates that a URL is a supported video source and optionally
    checks if the video is available for streaming.

    Returns compatibility issues for direct video URLs (codec checks, etc.).
    """
    try:
        # Normalize the URL
        normalized_url = SourceDetector.normalize_url(request.url)

        # Detect source type
        detection_result = SourceDetector.detect_source(normalized_url)

        if not detection_result["valid"]:
            return VideoSourceValidateResponse(
                valid=False,
                source_type=SourceType.UNKNOWN,
                source_type_label="Unknown",
                error=detection_result.get("error", "Unable to detect source type")
            )

        source_type = detection_result["source_type"]
        source_type_label = _get_source_type_label(source_type)

        # Check availability if requested (placeholder for future implementation)
        is_available = None
        compatibility_issues = []

        if request.check_availability:
            # TODO: Implement availability checking
            # This would involve making HTTP requests to validate the URL
            # and checking if the video is accessible
            pass

        # Check for compatibility issues
        if source_type == SourceType.DIRECT:
            # Direct video URLs may have codec compatibility issues
            metadata = detection_result.get("metadata", {})
            extension = metadata.get("extension", "")

            # Check for potentially problematic formats
            if extension in ['.avi', '.wmv', '.flv']:
                compatibility_issues.append(
                    f"File format {extension} may require transcoding for streaming"
                )

        return VideoSourceValidateResponse(
            valid=True,
            source_type=source_type,
            source_type_label=source_type_label,
            is_available=is_available,
            compatibility_issues=compatibility_issues,
            error=None
        )

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error validating video source: {str(e)}"
        )


@router.get("/supported", response_model=SupportedSourcesResponse)
async def get_supported_sources(
    current_user: User = Depends(get_current_user)
):
    """
    Get list of supported video source types.

    Returns information about all supported video source types,
    including descriptions and examples.
    """
    sources = [
        {
            "type": SourceType.YOUTUBE,
            "label": "YouTube",
            "description": "YouTube videos, playlists, and channels",
            "examples": [
                "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
                "https://youtu.be/dQw4w9WgXcQ",
                "https://www.youtube.com/playlist?list=PLxyz"
            ]
        },
        {
            "type": SourceType.VIMEO,
            "label": "Vimeo",
            "description": "Vimeo videos and playlists",
            "examples": [
                "https://vimeo.com/123456789",
                "https://vimeo.com/channels/staffpicks/123456789"
            ]
        },
        {
            "type": SourceType.TWITCH,
            "label": "Twitch",
            "description": "Twitch channels, VODs, and clips",
            "examples": [
                "https://www.twitch.tv/username",
                "https://www.twitch.tv/videos/123456789"
            ]
        },
        {
            "type": SourceType.DAILYMOTION,
            "label": "Dailymotion",
            "description": "Dailymotion videos",
            "examples": [
                "https://www.dailymotion.com/video/x123abc",
                "https://dai.ly/x123abc"
            ]
        },
        {
            "type": SourceType.DIRECT,
            "label": "Direct Video URL",
            "description": "Direct video file URLs (MP4, WebM, MKV, etc.)",
            "examples": [
                "https://example.com/videos/video.mp4",
                "https://cdn.example.com/content/video.webm"
            ]
        },
        {
            "type": SourceType.HLS,
            "label": "HLS Stream",
            "description": "HTTP Live Streaming (.m3u8)",
            "examples": [
                "https://example.com/stream/playlist.m3u8"
            ]
        },
        {
            "type": SourceType.DASH,
            "label": "DASH Stream",
            "description": "Dynamic Adaptive Streaming over HTTP (.mpd)",
            "examples": [
                "https://example.com/stream/manifest.mpd"
            ]
        },
        {
            "type": SourceType.GOOGLE_DRIVE,
            "label": "Google Drive",
            "description": "Videos stored on Google Drive",
            "examples": [
                "https://drive.google.com/file/d/1Ab2Cd3Ef4Gh5Ij6Kl7Mn8Op9Qr0St1U/view"
            ]
        },
        {
            "type": SourceType.DROPBOX,
            "label": "Dropbox",
            "description": "Videos stored on Dropbox",
            "examples": [
                "https://www.dropbox.com/s/abc123def456/video.mp4?dl=0"
            ]
        },
        {
            "type": SourceType.ONEDRIVE,
            "label": "OneDrive",
            "description": "Videos stored on Microsoft OneDrive",
            "examples": [
                "https://onedrive.live.com/?authkey=...",
                "https://1drv.ms/u/s!AbC123DeF456"
            ]
        },
        {
            "type": SourceType.RSS_FEED,
            "label": "RSS/Atom Feed",
            "description": "RSS or Atom feeds with video enclosures",
            "examples": [
                "https://example.com/feed.xml",
                "https://example.com/rss",
                "https://example.com/atom.xml"
            ]
        }
    ]

    return SupportedSourcesResponse(
        sources=sources,
        total_count=len(sources)
    )


# ============================================================================
# Helper Functions
# ============================================================================

def _get_source_type_label(source_type: SourceType) -> str:
    """
    Get human-readable label for source type.

    Args:
        source_type: SourceType enum value

    Returns:
        str: Human-readable label
    """
    labels = {
        SourceType.YOUTUBE: "YouTube",
        SourceType.VIMEO: "Vimeo",
        SourceType.DAILYMOTION: "Dailymotion",
        SourceType.TWITCH: "Twitch",
        SourceType.DIRECT: "Direct Video URL",
        SourceType.HLS: "HLS Stream",
        SourceType.DASH: "DASH Stream",
        SourceType.GOOGLE_DRIVE: "Google Drive",
        SourceType.DROPBOX: "Dropbox",
        SourceType.ONEDRIVE: "OneDrive",
        SourceType.RSS_FEED: "RSS/Atom Feed",
        SourceType.UNKNOWN: "Unknown"
    }

    return labels.get(source_type, str(source_type))
